From 25dd51e921191053b4cbeaf17306c25d91cf110b Mon Sep 17 00:00:00 2001
From: Ayush Dey <deyayush6@gmail.com>
Date: Mon, 11 Aug 2025 11:42:39 +0530
Subject: [PATCH 14/27] lib: add libvlc_downloader API

Used to download a file readable by VLC over a limited
set of protocols: http(s), ftp, file, nfs, smb, sftp

Refs #27971
---
 NEWS                            |   2 +
 include/meson.build             |   1 +
 include/vlc/libvlc_downloader.h | 362 ++++++++++++++++++++
 include/vlc/vlc.h               |   1 +
 lib/Makefile.am                 |   2 +
 lib/downloader.c                | 564 ++++++++++++++++++++++++++++++++
 lib/libvlc.sym                  |   7 +
 lib/meson.build                 |   1 +
 8 files changed, 940 insertions(+)
 create mode 100644 include/vlc/libvlc_downloader.h
 create mode 100644 lib/downloader.c

diff --git a/NEWS b/NEWS
index d01caed7c1..c62b1d0328 100644
--- a/NEWS
+++ b/NEWS
@@ -346,6 +346,8 @@ LibVLC:
  * Add libvlc_picture and libvlc_picture_list APIs, including RGBA and WebP
    support
  * Rework media parsing and thumbnailing around the new libvlc_parser API
+ * Add libvlc_downloader API that provides a bitstream of media to client
+   applications
  * Add media track list APIs, refcounted media tracks, multiple track
    selection and selected-track reporting
  * Add media-player timer/watch APIs, record API, program API, next-media API,
diff --git a/include/meson.build b/include/meson.build
index 9e9d4d5165..e022df2f75 100644
--- a/include/meson.build
+++ b/include/meson.build
@@ -3,6 +3,7 @@ install_headers(
     'vlc/vlc.h',
     'vlc/libvlc.h',
     'vlc/libvlc_dialog.h',
+    'vlc/libvlc_downloader.h',
     'vlc/libvlc_media.h',
     'vlc/libvlc_media_discoverer.h',
     'vlc/libvlc_media_list.h',
diff --git a/include/vlc/libvlc_downloader.h b/include/vlc/libvlc_downloader.h
new file mode 100644
index 0000000000..43a6343da6
--- /dev/null
+++ b/include/vlc/libvlc_downloader.h
@@ -0,0 +1,362 @@
+/*****************************************************************************
+ * libvlc_downloader.h:  LibVLC Downloader API
+ *****************************************************************************
+ * Copyright (C) 2026 VLC authors and VideoLAN
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
+ *****************************************************************************/
+
+#ifndef VLC_LIBVLC_DOWNLOADER_H
+#define VLC_LIBVLC_DOWNLOADER_H 1
+
+#include <vlc/libvlc.h>
+#include <vlc/libvlc_media.h>
+#include <vlc/libvlc_media_list.h>
+
+# ifdef __cplusplus
+extern "C" {
+# endif
+
+/** \defgroup libvlc_downloader LibVLC downloader
+ * \ingroup libvlc
+ * @ref libvlc_downloader_t is an abstract representation of a downloader
+ * @{
+ * \file
+ * LibVLC downloader API
+ */
+typedef struct libvlc_downloader_t libvlc_downloader_t;
+
+/**
+ * A downloader request object
+ */
+typedef struct libvlc_downloader_request_t libvlc_downloader_request_t;
+
+/**
+ * Opaque handle of a downloader task.
+ *
+ * Identifies a task request submitted via libvlc_downloader_queue().
+ * It can be passed to libvlc_downloader_cancel() to cancel that request,
+ * or to libvlc_downloader_set_pause() to pause/resume that request.
+ *
+ * \note Validity starts when libvlc_downloader_queue() returns a non-NULL handle
+ * and ends with libvlc_downloader_task_release().
+ */
+typedef struct libvlc_downloader_task libvlc_downloader_task;
+
+/**
+ * Downloader status
+ */
+typedef enum libvlc_downloader_status_t
+{
+    /** download pending */
+    libvlc_downloader_status_pending,
+    /** active download in progress (not paused) */
+    libvlc_downloader_status_running,
+    /** download paused */
+    libvlc_downloader_status_paused,
+    /** download finished */
+    libvlc_downloader_status_finished,
+    /** download cancelled */
+    libvlc_downloader_status_cancelled,
+    /** download error occurred */
+    libvlc_downloader_status_error,
+} libvlc_downloader_status_t;
+
+/** Sentinel return value to signal an error in the download (to be returned by on_buffer callback) */
+#define LIBVLC_DOWNLOADER_CB_ERROR ((ptrdiff_t)-1)
+
+/** Sentinel return value to cancel the download (to be returned by on_buffer callback) */
+#define LIBVLC_DOWNLOADER_CB_CANCEL ((ptrdiff_t)-2)
+
+/**
+ * Downloader callbacks
+ */
+struct libvlc_downloader_cbs
+{
+    /**
+     * Version of struct libvlc_downloader_cbs
+     */
+    uint32_t version;
+
+    /**
+     * Called when a buffer of data is read.
+     *
+     * \note Mandatory (can't be NULL),
+     * available since version 0
+     *
+     * \warning Do not call any libvlc_downloader_* API from within this callback.
+     * Only libvlc_downloader_task_* APIs may be called on the provided task handle.
+     * And avoid blocking operations in this callback as it is invoked with the internal lock held.
+     *
+     * \param opaque user data
+     * \param task opaque handle returned by libvlc_downloader_queue()
+     * \param buf pointer to buffer (owned by downloader, only valid during callback)
+     * \param len size of buffer
+     * \param position total number of bytes read by the downloader so far
+     * \param total total size of the media in bytes (only medias with finite size are allowed to download)
+     *
+     * \return The number of bytes consumed by the user, or \ref LIBVLC_DOWNLOADER_CB_ERROR to
+     * terminate the download with an error, or \ref LIBVLC_DOWNLOADER_CB_CANCEL to cancel the download.
+     *
+     * If the returned number of bytes consumed is less than \p len (partial read), the downloader
+     * auto-pauses as a form of backpressure. The user must call `libvlc_downloader_set_pause(dl, task, false)`
+     * from the main thread to resume when ready to consume data again.
+     *
+     * \note User must copy the data if it needs to be accessed later.
+     * The maximum buffer size is limited to around 64 KB.
+     */
+    ptrdiff_t (*on_buffer)(void *opaque, libvlc_downloader_task *task,
+                           const uint8_t *buf, size_t len,
+                           uint64_t position, uint64_t total);
+
+    /**
+     * Called when the downloader state changes.
+     *
+     * \note Mandatory (can't be NULL),
+     * available since version 0
+     *
+     * Invoked when download starts, pauses, resumes, cancels, finishes
+     * or errors out. \ref libvlc_downloader_status_t
+     *
+     * \warning Do not call any libvlc_downloader_* API from within this callback.
+     * Only libvlc_downloader_task_* APIs may be called on the provided task handle.
+     * And avoid blocking operations in this callback as it is invoked with the internal lock held.
+     *
+     * \param opaque user data
+     * \param task opaque handle returned by libvlc_downloader_queue()
+     * \param status download status
+     */
+    void (*on_state_update)(void *opaque, libvlc_downloader_task *task,
+                            libvlc_downloader_status_t status);
+
+    /**
+     * Called when subitems of the media are available.
+     *
+     * \note Optional (can be NULL),
+     * available since version 0
+     *
+     * \param opaque user data
+     * \param task opaque handle returned by libvlc_downloader_queue()
+     * \param subitems media list of subitems (owned by LibVLC)
+     */
+    void (*on_subitems)(void *opaque, libvlc_downloader_task *task,
+                        libvlc_media_list_t *subitems);
+
+    /**
+     * Called when the parsed media has slaves.
+     *
+     * \note Optional (can be NULL),
+     * available since version 0
+     *
+     * \param opaque user data
+     * \param task opaque handle returned by libvlc_downloader_queue()
+     * \param slaves array of libvlc_media_slave_t* (owned by LibVLC)
+     * \param count number of slaves
+     */
+    void (*on_slaves)(void *opaque, libvlc_downloader_task *task,
+                      libvlc_media_slave_t **slaves, size_t count);
+};
+
+/**
+ * struct defining a downloader request
+ */
+struct libvlc_downloader_request_t
+{
+    /**
+     * Version of libvlc_downloader_request_t
+     */
+    uint32_t version;
+
+    /**
+     * Media to download
+     *
+     * \note Mandatory (can't be NULL),
+     * available since version 0
+     *
+     * - Only finite-size media are allowed to download.
+     *
+     * - If the media is a playlist or directory, the user will be notified of the
+     *   subitems via the on_subitems callback and the download will not proceed.
+     *
+     * - If the media is a livestream or unknown type, the download will error out.
+     */
+    libvlc_media_t *media;
+};
+
+/**
+ * struct defining downloader configuration
+ */
+struct libvlc_downloader_cfg
+{
+    /**
+     * Version of struct libvlc_downloader_cfg
+     */
+    uint32_t version;
+
+    /**
+     * The maximum number of threads used by the parser internally, 0 for default
+     * (1 thread)
+     *
+     * \note Optional (can be 0),
+     * available since version 0
+     */
+    uint32_t max_parser_threads;
+};
+
+/**
+ * Create a downloader instance.
+ *
+ * Supports downloading files over a limited set of protocols:
+ * http(s), ftp, file, nfs, smb, sftp
+ *
+ * The downloader must be released by calling libvlc_downloader_destroy()
+ * when it is no longer needed.
+ *
+ * \param inst LibVLC instance
+ * \param cfg a pointer to a valid downloader configuration struct
+ * \return downloader instance or NULL on error
+ * 
+ * \version LibVLC 4.0.0 or later
+ */
+LIBVLC_API libvlc_downloader_t *
+libvlc_downloader_new(libvlc_instance_t *inst, const struct libvlc_downloader_cfg *cfg);
+
+/**
+ * Download a media asynchronously.
+ *
+ * - The downloader first parses the media.
+ *
+ * - If the media has subitems, the user will be notified via the on_subitems callback.
+ *
+ * - If the media has slaves, the user will be notified via the on_slaves callback.
+ *
+ * - If the media is not a file type,
+ *   the download will not proceed. \see libvlc_media_type_t,
+ *   and the user will be notified via the on_state_update callback.
+ *
+ * - If the media is a file type with finite size, the download starts in a separate thread.
+ *
+ * \param downloader downloader instance
+ * \param req a pointer to a valid request struct
+ * \param cbs a pointer to a valid callbacks struct. The pointed struct
+ * must be kept alive (and not modified) by the caller until libvlc_downloader_cbs.on_state_update()
+ * is called for the returned task handle with a terminal state (finished/cancelled/error).
+ * \param cbs_opaque opaque pointer for callbacks
+ * \return NULL in case of error, or a valid handle if the request was
+ * scheduled for downloading.
+ *
+ * \note No callbacks will be invoked if the return value is NULL.
+ *
+ * \version LibVLC 4.0.0 or later
+ */
+LIBVLC_API libvlc_downloader_task *
+libvlc_downloader_queue(libvlc_downloader_t *downloader, const libvlc_downloader_request_t *req,
+                        const struct libvlc_downloader_cbs *cbs, void *cbs_opaque);
+
+/**
+ * Cancel an ongoing download.
+ *
+ * \param downloader downloader instance
+ * \param task a downloader task returned by libvlc_downloader_queue(), 
+ * or NULL to cancel all requests.
+ *
+ * \return the number of requests cancelled
+ *
+ * \note
+ * - This function is valid only if the request is in one of the following states:
+ *   pending, running, or paused.
+ *
+ * - When a request is cancelled, the `on_state_update` callback will be triggered
+ *   with the cancelled state.
+ *
+ * - If the request is already in a terminated state (finished, cancelled, or error),
+ *   the call is a no-op and no callback will be invoked.
+ *
+ * \version LibVLC 4.0.0 or later
+ */
+LIBVLC_API size_t libvlc_downloader_cancel(libvlc_downloader_t *downloader, libvlc_downloader_task *task);
+
+/**
+ * Toggle pause/resume for the download.
+ *
+ * \param downloader downloader instance
+ * \param task a valid downloader task returned by libvlc_downloader_queue()
+ * \param paused true to pause, false to resume
+ *
+ * \note This API is valid only when the download is in pending/running/paused state.
+ * And the on_state_update callback with paused/running state will be called only during these
+ * state changes. Else, for finished/cancelled/error states, it's a no-op and
+ * no callback will be called.
+ *
+ * \version LibVLC 4.0.0 or later
+ */
+LIBVLC_API void libvlc_downloader_set_pause(libvlc_downloader_t *downloader,
+                                            libvlc_downloader_task *task,
+                                            bool paused);
+
+/**
+ * Destroy a downloader and free resources.
+ * All pending, running and paused downloads are cancelled.
+ * Waits for all download threads to join.
+ *
+ * \param downloader downloader instance
+ *
+ * \version LibVLC 4.0.0 or later
+ */
+LIBVLC_API void libvlc_downloader_destroy(libvlc_downloader_t *downloader);
+
+/**
+ * Get the media associated with the downloader request handle.
+ *
+ * \param task opaque handle returned by libvlc_downloader_queue()
+ * \return the media associated with the request handle.
+ *
+ * \note The returned media is held by the task, it must not be
+ * released by the caller.
+ *
+ * \version LibVLC 4.0.0 or later
+ */
+LIBVLC_API libvlc_media_t *
+libvlc_downloader_task_get_media(libvlc_downloader_task *task);
+
+/**
+ * Release a downloader task handle.
+ *
+ * \param task the downloader task handle
+ *
+ * \note
+ * - The task handle is retained when returned by libvlc_downloader_queue().
+ *
+ * - Mandatory to call to avoid memory leaks.
+ *
+ * - It is safe to call this API from within the on_state_update callback, when it
+ *   reports a terminal state (finished, cancelled, error) \see libvlc_downloader_status_t.
+ *
+ * - The task handle should not be used after calling this function.
+ *
+ * - If called on an active task, it doesn't cancel the task,
+ *   use \ref libvlc_downloader_cancel() for that.
+ *
+ * \version LibVLC 4.0.0 or later
+ */
+LIBVLC_API void libvlc_downloader_task_release(libvlc_downloader_task *task);
+
+/** @}*/
+
+# ifdef __cplusplus
+}
+# endif
+
+#endif /* VLC_LIBVLC_DOWNLOADER_H */
diff --git a/include/vlc/vlc.h b/include/vlc/vlc.h
index 67c894a330..5b6f1d9ab0 100644
--- a/include/vlc/vlc.h
+++ b/include/vlc/vlc.h
@@ -45,6 +45,7 @@ extern "C" {
 #include "libvlc_media_list_player.h"
 #include "libvlc_media_discoverer.h"
 #include "libvlc_dialog.h"
+#include "libvlc_downloader.h"
 #include "libvlc_version.h"
 
 # ifdef __cplusplus
diff --git a/lib/Makefile.am b/lib/Makefile.am
index 1592458c57..cda74c8037 100644
--- a/lib/Makefile.am
+++ b/lib/Makefile.am
@@ -10,6 +10,7 @@ CLEANFILES = $(BUILT_SOURCES) $(pkgconfig_DATA)
 pkginclude_HEADERS = \
 	../include/vlc/libvlc.h \
 	../include/vlc/libvlc_dialog.h \
+	../include/vlc/libvlc_downloader.h \
 	../include/vlc/libvlc_media.h \
 	../include/vlc/libvlc_media_discoverer.h \
 	../include/vlc/libvlc_media_list.h \
@@ -34,6 +35,7 @@ libvlc_la_SOURCES = \
 	renderer_discoverer_internal.h \
 	core.c \
 	dialog.c \
+	downloader.c \
 	renderer_discoverer.c \
 	error.c \
 	log.c \
diff --git a/lib/downloader.c b/lib/downloader.c
new file mode 100644
index 0000000000..6431fad856
--- /dev/null
+++ b/lib/downloader.c
@@ -0,0 +1,564 @@
+/*****************************************************************************
+ * downloader.c: LibVLC Downloader API
+ *****************************************************************************
+ * Copyright (C) 2026 VLC authors and VideoLAN
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
+ *****************************************************************************/
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <vlc/libvlc_downloader.h>
+
+#include <vlc_access.h>
+#include <vlc_interrupt.h>
+#include <vlc_stream.h>
+#include <vlc_threads.h>
+
+#include "libvlc_internal.h"
+
+#define BUFFER_SIZE 65536
+
+struct libvlc_downloader_thread
+{
+    /* Node of libvlc_downloader_t.threads list */
+    struct vlc_list node;
+
+    /* The task thread */
+    vlc_thread_t thread;
+
+    /* Thread termination state */
+    bool terminated;
+};
+
+struct libvlc_downloader_t
+{
+    libvlc_instance_t *libvlc;
+    vlc_mutex_t lock;
+    libvlc_parser_t *parser;
+
+    /* list of ongoing tasks (terminated tasks are removed) */
+    struct vlc_list submitted_tasks;
+
+    /* list of downloader threads (dead threads joined at libvlc_downloader_queue,
+       all threads are joined at libvlc_downloader_destroy) */
+    struct vlc_list threads;
+};
+
+struct libvlc_downloader_task
+{
+    libvlc_downloader_t *downloader;
+    struct vlc_list node;
+    libvlc_downloader_status_t status;
+
+    libvlc_parser_task *parser_task;
+
+    libvlc_media_t *media;
+    stream_t *s;
+
+    const struct libvlc_downloader_cbs *cbs;
+    void *cbs_opaque;
+
+    /* interrupt context (to interrupt blocking read) */
+    vlc_interrupt_t *interrupt;
+
+    /* required for pause/resume */
+    bool pause_requested;
+    vlc_cond_t interrupt_cond;
+
+    /* required for ongoing task abortion */
+    bool interrupted;
+
+    vlc_atomic_rc_t rc;
+
+    struct libvlc_downloader_thread *thread;
+};
+
+/* Create a new downloader task */
+static struct libvlc_downloader_task *
+DownloaderTaskNew(libvlc_downloader_t *downloader, libvlc_media_t *media,
+                  const struct libvlc_downloader_cbs *cbs, void *cbs_opaque)
+{
+    struct libvlc_downloader_task *task = malloc(sizeof(*task));
+    if (task == NULL)
+        return NULL;
+
+    task->interrupt = vlc_interrupt_create();
+    if (task->interrupt == NULL)
+    {
+        free(task);
+        return NULL;
+    }
+
+    task->downloader = downloader;
+    task->media = libvlc_media_retain(media);
+    task->s = NULL;
+    task->cbs = cbs;
+    task->cbs_opaque = cbs_opaque;
+    task->pause_requested = false;
+    vlc_cond_init(&task->interrupt_cond);
+    task->interrupted = false;
+    task->status = libvlc_downloader_status_pending;
+    task->parser_task = NULL;
+    task->thread = NULL;
+
+    vlc_atomic_rc_init(&task->rc);
+
+    return task;
+}
+
+/* Delete a downloader task */
+static void
+DownloaderTaskDestroy(struct libvlc_downloader_task *task)
+{
+    libvlc_media_release(task->media);
+    vlc_interrupt_destroy(task->interrupt);
+
+    if (task->parser_task)
+        libvlc_parser_task_release(task->parser_task);
+
+    if (task->s)
+        vlc_stream_Delete(task->s);
+
+    free(task);
+}
+
+/* Interrupt a running downloader task */
+static void
+DownloaderTaskInterruptLocked(struct libvlc_downloader_task *task)
+{
+    task->interrupted = true;
+    vlc_interrupt_kill(task->interrupt);
+    vlc_cond_signal(&task->interrupt_cond);
+}
+
+/* report downloader state changes */
+static void libvlc_downloader_task_ReportStateLocked(struct libvlc_downloader_task *task,
+                                                     libvlc_downloader_status_t new_status)
+{
+    bool status_changed = false;
+
+    if (task->interrupted)
+        new_status = libvlc_downloader_status_cancelled;
+
+    libvlc_downloader_status_t old_status = task->status;
+    if (old_status != new_status)
+    {
+        task->status = new_status;
+        status_changed = true;
+    }
+
+    /* report on_state_update only when the state changed */
+    if (status_changed)
+        task->cbs->on_state_update(task->cbs_opaque, task, new_status);
+}
+
+static void *
+downloaderThread(void *data)
+{
+    vlc_thread_set_name("downloader");
+    struct libvlc_downloader_task *task = data;
+    libvlc_downloader_t *downloader = task->downloader;
+    libvlc_downloader_status_t status = libvlc_downloader_status_error;
+    uint8_t *buf = NULL;
+
+    vlc_interrupt_set(task->interrupt);
+
+    char *mrl = libvlc_media_get_mrl(task->media);
+    if (mrl == NULL)
+        goto cleanup;
+
+    task->s = vlc_access_NewMRL(VLC_OBJECT(downloader->libvlc->p_libvlc_int), mrl);
+    free(mrl);
+
+    if (task->s == NULL || !vlc_stream_CanDownload(task->s))
+        goto cleanup;
+
+    uint64_t stream_size = 0;
+    /* only files with valid size allowed */
+    if (vlc_stream_GetSize(task->s, &stream_size) != VLC_SUCCESS)
+        goto cleanup;
+
+    buf = malloc(BUFFER_SIZE);
+    if (buf == NULL)
+        goto cleanup;
+
+    uint64_t total_read = 0;
+    size_t buf_len = 0;
+    size_t consumed_offset = 0;
+    status = libvlc_downloader_status_cancelled;
+
+    while (true)
+    {
+        vlc_mutex_lock(&downloader->lock);
+        if (task->interrupted)
+            goto cleanup_locked;
+
+        /* Pause handling */
+        if (task->pause_requested)
+        {
+            libvlc_downloader_task_ReportStateLocked(task, libvlc_downloader_status_paused);
+            vlc_cond_wait(&task->interrupt_cond, &downloader->lock);
+            vlc_mutex_unlock(&downloader->lock);
+            continue;
+        }
+
+        libvlc_downloader_task_ReportStateLocked(task, libvlc_downloader_status_running);
+
+        /* fetch next chunk only when the previous one is fully drained */
+        if (consumed_offset == buf_len)
+        {
+            vlc_mutex_unlock(&downloader->lock);
+
+            ssize_t read = vlc_stream_ReadPartial(task->s, buf, BUFFER_SIZE);
+
+            if (read < 0)
+                continue;
+
+            vlc_mutex_lock(&downloader->lock);
+            /* check whether task was interrupted during blocking read operation */
+            if (task->interrupted)
+                goto cleanup_locked;
+
+            if (read == 0) /* EOF */
+            {
+                status = libvlc_downloader_status_finished;
+                goto cleanup_locked;
+            }
+
+            buf_len = (size_t)read;
+            consumed_offset = 0;
+            total_read += (uint64_t)read;
+
+            /* update size in case it changed; a valid size is required throughout */
+            if (vlc_stream_GetSize(task->s, &stream_size) != VLC_SUCCESS)
+            {
+                status = libvlc_downloader_status_error;
+                goto cleanup_locked;
+            }
+
+            /* check whether the user requested pause during the blocking read. In that case,
+               the data is buffered and download is paused. */
+            if (task->pause_requested)
+            {
+                vlc_mutex_unlock(&downloader->lock);
+                continue;
+            }
+        }
+
+        /* deliver the residual (either just fetched or left over after a partial read) */
+        const uint8_t *chunk = buf + consumed_offset;
+        size_t remaining = buf_len - consumed_offset;
+
+        ptrdiff_t consumed = task->cbs->on_buffer(task->cbs_opaque, task, chunk,
+                                                  remaining, total_read, stream_size);
+
+        if (consumed == LIBVLC_DOWNLOADER_CB_CANCEL)
+            goto cleanup_locked;
+
+        if (consumed < 0 || (size_t)consumed > remaining)
+        {
+            status = libvlc_downloader_status_error;
+            goto cleanup_locked;
+        }
+
+        consumed_offset += (size_t)consumed;
+
+        /* partial read: auto-pause as backpressure */
+        if ((size_t)consumed < remaining)
+            task->pause_requested = true;
+
+        vlc_mutex_unlock(&downloader->lock);
+    }
+
+cleanup:
+    vlc_mutex_lock(&downloader->lock);
+cleanup_locked:
+    libvlc_downloader_task_ReportStateLocked(task, status);
+    vlc_list_remove(&task->node);
+    task->thread->terminated = true;
+    vlc_mutex_unlock(&downloader->lock);
+    free(buf);
+    libvlc_downloader_task_release(task);
+    return NULL;
+}
+
+libvlc_downloader_t *
+libvlc_downloader_new(libvlc_instance_t *inst, const struct libvlc_downloader_cfg *cfg)
+{
+    assert(inst != NULL);
+    assert(cfg != NULL);
+
+    /* No different versions to handle for now */
+    assert(cfg->version <= 0);
+
+    libvlc_downloader_t *downloader = malloc(sizeof(*downloader));
+    if (downloader == NULL)
+        return NULL;
+
+    const struct libvlc_parser_cfg parser_cfg = {
+        .version = 0,
+        .max_parser_threads = cfg->max_parser_threads,
+        .timeout = 0,
+    };
+
+    downloader->parser = libvlc_parser_new(inst, &parser_cfg);
+    if (downloader->parser == NULL)
+    {
+        free(downloader);
+        return NULL;
+    }
+
+    downloader->libvlc = libvlc_retain(inst);
+    vlc_mutex_init(&downloader->lock);
+    vlc_list_init(&downloader->submitted_tasks);
+    vlc_list_init(&downloader->threads);
+    return downloader;
+}
+
+static void notify_subitems(const struct libvlc_downloader_cbs *cbs, void *cbs_opaque,
+                            libvlc_downloader_task *task)
+{
+    libvlc_media_list_t *mlist = libvlc_media_subitems(task->media);
+    if (mlist == NULL)
+        return;
+
+    cbs->on_subitems(cbs_opaque, task, mlist);
+    libvlc_media_list_release(mlist);
+}
+
+static void notify_slaves(const struct libvlc_downloader_cbs *cbs, void *cbs_opaque,
+                          libvlc_downloader_task *task)
+{
+    libvlc_media_slave_t **slaves = NULL;
+    unsigned count = libvlc_media_slaves_get(task->media, &slaves);
+    if (count > 0 && slaves)
+    {
+        cbs->on_slaves(cbs_opaque, task, slaves, count);
+        libvlc_media_slaves_release(slaves, count);
+    }
+}
+
+static void parse_ended(void *opaque, libvlc_parser_task *parser_task, libvlc_parser_status_t status)
+{
+    (void) parser_task;
+    struct libvlc_downloader_task *task = opaque;
+    const struct libvlc_downloader_cbs *cbs = task->cbs;
+    void *cbs_opaque = task->cbs_opaque;
+    libvlc_downloader_t *downloader = task->downloader;
+    libvlc_downloader_status_t downloader_status = libvlc_downloader_status_error;
+
+    if (status == libvlc_parser_status_cancelled)
+    {
+        downloader_status = libvlc_downloader_status_cancelled;
+        goto error;
+    }
+
+    if (status != libvlc_parser_status_done)
+        goto error;
+
+    if (cbs->on_subitems)
+        notify_subitems(cbs, cbs_opaque, task);
+
+    if (cbs->on_slaves)
+        notify_slaves(cbs, cbs_opaque, task);
+
+    libvlc_media_type_t type = libvlc_media_get_type(task->media);
+    if (type != libvlc_media_type_file)
+        goto error;
+
+    struct libvlc_downloader_thread *thread = malloc(sizeof(*thread));
+    if (thread == NULL)
+        goto error;
+
+    task->thread = thread;
+
+    vlc_mutex_lock(&downloader->lock);
+    thread->terminated = false;
+    vlc_list_append(&thread->node, &downloader->threads);
+    if (vlc_clone(&thread->thread, downloaderThread, task))
+    {
+        vlc_list_remove(&thread->node);
+        vlc_mutex_unlock(&downloader->lock);
+        free(thread);
+        task->thread = NULL;
+        goto error;
+    }
+    vlc_mutex_unlock(&downloader->lock);
+
+    return;
+
+error:
+    vlc_mutex_lock(&downloader->lock);
+    libvlc_downloader_task_ReportStateLocked(task, downloader_status);
+    vlc_list_remove(&task->node);
+    vlc_mutex_unlock(&downloader->lock);
+    libvlc_downloader_task_release(task);
+}
+
+static const struct libvlc_parser_cbs parser_cbs = {
+    .version = 0,
+    .on_parsed = parse_ended,
+    .on_attachments_added = NULL,
+};
+
+libvlc_downloader_task *
+libvlc_downloader_queue(libvlc_downloader_t *downloader, const libvlc_downloader_request_t *req,
+                        const struct libvlc_downloader_cbs *cbs, void *cbs_opaque)
+{
+    assert(downloader != NULL);
+    assert(req != NULL && req->media != NULL);
+    assert(cbs != NULL && cbs->on_buffer != NULL && cbs->on_state_update != NULL);
+
+    /* No different versions to handle for now */
+    assert(req->version <= 0);
+    assert(cbs->version <= 0);
+
+    struct libvlc_downloader_task *task = DownloaderTaskNew(downloader, req->media,
+                                                            cbs, cbs_opaque);
+
+    if (task == NULL)
+        return NULL;
+
+    vlc_mutex_lock(&downloader->lock);
+    vlc_list_append(&task->node, &downloader->submitted_tasks);
+
+    /* clean up terminated threads */
+    struct libvlc_downloader_thread *thread;
+    vlc_list_foreach(thread, &downloader->threads, node)
+    {
+        if (thread->terminated)
+        {
+            vlc_join(thread->thread, NULL);
+            vlc_list_remove(&thread->node);
+            free(thread);
+        }
+    }
+    vlc_mutex_unlock(&downloader->lock);
+
+    libvlc_parser_request_t request = {
+        .version = 0,
+        .media = req->media,
+        .parse_flags = 0, /* default VLC_PREPARSER_TYPE_PARSE | VLC_PREPARSER_OPTION_SUBITEMS */
+    };
+
+    vlc_atomic_rc_inc(&task->rc);
+
+    libvlc_parser_task *parser_task = libvlc_parser_queue(downloader->parser, &request,
+                                                          &parser_cbs, task);
+    if (parser_task == NULL)
+    {
+        vlc_mutex_lock(&downloader->lock);
+        vlc_list_remove(&task->node);
+        vlc_mutex_unlock(&downloader->lock);
+        DownloaderTaskDestroy(task);
+        return NULL;
+    }
+
+    task->parser_task = parser_task;
+    return task;
+}
+
+size_t libvlc_downloader_cancel(libvlc_downloader_t *downloader, libvlc_downloader_task *task)
+{
+    assert(downloader != NULL);
+    size_t cancelled = 0;
+
+    if (task != NULL)
+    {
+        cancelled = libvlc_parser_cancel_request(downloader->parser, task->parser_task);
+        if (cancelled > 0)
+            return cancelled;
+    }
+    else
+        cancelled = libvlc_parser_cancel_request(downloader->parser, NULL);
+
+    vlc_mutex_lock(&downloader->lock);
+
+    struct libvlc_downloader_task *task_itr;
+    vlc_list_foreach(task_itr, &downloader->submitted_tasks, node)
+    {
+        if (task != NULL && task_itr != task)
+            continue;
+
+        if (task_itr->thread != NULL)
+        {
+            cancelled++;
+            DownloaderTaskInterruptLocked(task_itr);
+        }
+    }
+
+    vlc_mutex_unlock(&downloader->lock);
+
+    return cancelled;
+}
+
+void libvlc_downloader_set_pause(libvlc_downloader_t *downloader, libvlc_downloader_task *task,
+                                 bool paused)
+{
+    assert(downloader != NULL);
+    assert(task != NULL);
+
+    vlc_mutex_lock(&downloader->lock);
+    task->pause_requested = paused;
+    if (!paused)
+        vlc_cond_signal(&task->interrupt_cond);
+    vlc_mutex_unlock(&downloader->lock);
+}
+
+void libvlc_downloader_destroy(libvlc_downloader_t *downloader)
+{
+    assert(downloader != NULL);
+    libvlc_downloader_cancel(downloader, NULL);
+
+    /* drain the parser first. libvlc_parser_destroy() waits for all in-flight
+       parse_ended callbacks to complete. After this returns, no task can still
+       be in the parse phase and no new download thread can be spawned, so the
+       threads list is stable and can be joined without racing the parser. */
+    libvlc_parser_destroy(downloader->parser);
+
+    struct libvlc_downloader_thread *thread;
+    vlc_list_foreach(thread, &downloader->threads, node)
+    {
+        vlc_join(thread->thread, NULL);
+        vlc_list_remove(&thread->node);
+        free(thread);
+    }
+
+    /* both lists must be empty by now */
+    assert(vlc_list_is_empty(&downloader->submitted_tasks));
+    assert(vlc_list_is_empty(&downloader->threads));
+
+    libvlc_release(downloader->libvlc);
+    free(downloader);
+}
+
+libvlc_media_t *
+libvlc_downloader_task_get_media(libvlc_downloader_task *task)
+{
+    assert(task != NULL);
+    return task->media;
+}
+
+void libvlc_downloader_task_release(libvlc_downloader_task *task)
+{
+    assert(task != NULL);
+    if (!vlc_atomic_rc_dec(&task->rc))
+        return;
+
+    DownloaderTaskDestroy(task);
+}
diff --git a/lib/libvlc.sym b/lib/libvlc.sym
index b6d66e2881..7b702ca6a7 100644
--- a/lib/libvlc.sym
+++ b/lib/libvlc.sym
@@ -119,6 +119,13 @@ libvlc_parser_queue_thumbnailing
 libvlc_parser_cancel_request
 libvlc_parser_task_get_media
 libvlc_parser_task_release
+libvlc_downloader_new
+libvlc_downloader_queue
+libvlc_downloader_cancel
+libvlc_downloader_set_pause
+libvlc_downloader_destroy
+libvlc_downloader_task_get_media
+libvlc_downloader_task_release
 libvlc_media_player_add_slave
 libvlc_media_player_can_pause
 libvlc_media_player_program_scrambled
diff --git a/lib/meson.build b/lib/meson.build
index 59ad21e6d3..8a29f2a618 100644
--- a/lib/meson.build
+++ b/lib/meson.build
@@ -1,6 +1,7 @@
 libvlc_sources = [
     'core.c',
     'dialog.c',
+    'downloader.c',
     'renderer_discoverer.c',
     'error.c',
     'log.c',
-- 
2.50.1 (Apple Git-155)

